enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,332 @@
+- Maintain a single code base/manifests for all the on-prem platforms.
+- Provide the same Kube-api downtime values ​​as in stack implemented only with static pods for master reboot and master drain cases.
+
+### Non-Goals
+
+- Define a new self-hosted stack for DNS or load balancing
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/on-prem) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by both MCO and the new operator.
+
+The following functionalities are expected as part of the new operator:
+
+- Introduce a CRD used by the operator to manage the self-hosted stack.
+- Fields in CR spec that control the self-hosted stack configuration.
+- Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+- Run Coredns-MDNS, HAProxy and MDNS-publisher self-hosted stack components as K8S DaemonSets (while Keepalived will continue running as static pod)
+
+## Risks and Mitigations
+
+- Impact on on-prem platforms
+- Upgrades/Downgrades
+- Disaster recovery
+- Cluster shutdown and resume/restart
+
+In order to mitigate the impact on on-prem platforms, in fresh deployments all the self-hosted LB, DNS, and VIP services will be enabled, therefore there is no need to rely on external services during deployment, also the on-prem platforms tests should be updated according to Test Plan section described below.
+
+Additionally, a detailed document in which the supported LB, DNS migration processes will be issued.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Is `cluster-hosted-net-services-operator` an acceptable name?
+- In which namespace the new operator components should be hosted?
+
+### current implmentation
+
+In current implementation, the [networking infra componenets](https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md) run as static pods through the MCO.
+
+### Suggested design
+
+To support early clustering requirements Keepalived will continue running as static pods through MCO, additionally, a new dnsmasq service (also deployed through MCO) will run in each node while HAProxy, CoreDNS-MDNS and MDNS-publisher will run as daemonset pods through the new operator.
+
+The following table reflects the suggested changes.
+
+| Componenet Name       | Current implementation | Proposed implementation |
+| --------------------- | ---------------------- | ----------------------- |
+| Resolv.conf prepender | NM dispatcher script   | NM dispatcher script    |
+| Keepalived            | Static pod             | Static pod              |
+| HAProxy               | Static pod             | DaemonSet pod           |
+| CoreDNS-MDNS          | Static pod             | DaemonSet pod           |
+| mDNS-Publisher        | Static pod             | DaemonSet pod           |
```

Comment:
Is DNS an highly-available service in the sense that one pod in the cluster is enough to answer all requests? Or do we need a running pod on every node?

Reply:
In case of self-hosted DNS, the CoreDNS-MDNS and MDNS-Publisher are used to resolve records that are automatically created using a cloud platform's DNS service in other platforms. see [1] for more details.

For cloud platforms we have the following  DNS servers hierarchy :
pod --> cluster-dns (provided by dns-operator) --> host's resolv.conf --> upstream servers

And this is DNS servers hierarchy  for on-prem:
 pod --> cluster-dns (provided by dns-operator) --> host's resolv.conf --> localhost:CoreDNS-MDNS --> upstream servers

We run MDNS-Publisher in each node (every node should advertise itself)
- For CoreDNS-MDNS we could:
    - Run (1+N) instances, assign a DNS-VIP to active node, and point  node's etc/resolv.conf to DNS-VIP address
    - Run instance in each node and point etc/resolv.conf to node's local IP address
    
We decided to go with the latter. 

BTW, with the approach proposed on the enhancement, unlike the current implementation, if the customer prefers to use his DNS server the CoreDNS-MDNS and MDNS-Publisher (that should stop the multicast traffic)  should be deleted from all nodes.


[1] https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md#internal-dns